File: src/core/WitchSaga.ts
import { People, Person } from "../models/index";
import { IsNegative, GetAverage } from "../utils/index";

class WitchSaga {
    private _people: People;

    constructor(people: People) {
        this._people = people;
    }

     /**
      * Check if person object is valid
      * @param person of type Person
      * @returns true or false
      */
    private IsPersonValid(person: Person): boolean {
        if (IsNegative(person.yearOfDeath) || IsNegative(person.ageOfDeath))
            return false;

        return person.yearOfDeath > person.ageOfDeath;
    }

    /**
     * Check if the given people object is valid
     * @returns true or false
     */
    private IsValid(): boolean {
        if (!this._people.personA || !this._people.personB || !this._people.personA.ageOfDeath || !this._people.personA.yearOfDeath || !this._people.personB.ageOfDeath || !this._people.personB.yearOfDeath)
            return false;

        if (!this.IsPersonValid(this._people.personA)
            || !this.IsPersonValid(this._people.personB))
            return false;

        return true;
    }

    /**
     * Get the number of people killed based on a given year
     * @param year 
     * @returns the number of people killed
     */
    private GetNumberOfPeopleKilled(year: number): number {
        //pattern: after the first and second, add the two previous ones plus 1
        // 1 + 2 + 1 = 4
        // 2 + 4 + 1 = 7
        // 4 + 7 + 1 = 12 
        // etc
        //[1, 2, 4, 7, 12, 20, 33, 54, 88 ...]
        let array = [];
        
        for (let x = 1; x <= year; x++) {
            if (x == 1 || x == 2) //no calculation if 1 or 2
                array.push(x)
            else {
                //add two last ones from the array + 1
                array.push(array[array.length - 1] + array[array.length - 2] + 1);
            }

        }

        return array[year - 1];
    }

    /**
     * Get the average number of people killed by the witch
     * @returns average number
     */
    public FindAverageOfPeopleKilled(): number {
        if (!this.IsValid())
            return -1;
    
        var yearOfPersonA = this._people.personA.yearOfDeath - this._people.personA.ageOfDeath;
        var yearOfPersonB = this._people.personB.yearOfDeath - this._people.personB.ageOfDeath;
    
        console.log("person A:" + this.GetNumberOfPeopleKilled(yearOfPersonA));
        console.log("person B:" + this.GetNumberOfPeopleKilled(yearOfPersonB));
    
        console.log("Average: " + GetAverage(this.GetNumberOfPeopleKilled(yearOfPersonA), this.GetNumberOfPeopleKilled(yearOfPersonB)));
        

        const average = GetAverage(this.GetNumberOfPeopleKilled(yearOfPersonA), this.GetNumberOfPeopleKilled(yearOfPersonB));

        return average;
    }
}

export default WitchSaga;

File: dotnet/WitchSaga/Core/WitchSaga.cs
using System.Collections.Generic;
using WitchSaga.Extensions;
using WitchSaga.Models;
using WitchSaga.Utilities;

namespace WitchSaga.Core
{
    public class WitchSaga
    {
        private People _people;
        public WitchSaga(People people)
        {
            _people = people;
        }

        /// <summary>
        /// Get the number of people killed based on a given year
        /// </summary>
        /// <param name="year"></param>
        /// <returns></returns>
        private double GetNumberOfPeopleKilled(int year)
        {
            //pattern: after the first and second, add the two previous ones plus 1
            // 1 + 2 + 1 = 4
            // 2 + 4 + 1 = 7
            // 4 + 7 + 1 = 12 
            // etc
            //[1, 2, 4, 7, 12, 20, 33, 54, 88 ...]
            var list = new List<int>();

            for (int x = 1; x <= year; x++)
            {
                if (x == 1 || x == 2) //no calculation if 1 or 2
                    list.Add(x);
                else
                {
                    //add two last ones from the array + 1
                    list.Add(list[list.Count - 1] + list[list.Count - 2] + 1);
                }
            }

            return list[year - 1];
        }

        /// <summary>
        /// Get the average number of people killed by the witch
        /// </summary>
        /// <returns></returns>
        public double FindAverageOfPeopleKilled()
        {
            if (!_people.IsValid())
                return -1;

            int yearOfPersonA = _people.PersonA.YearOfDeath - _people.PersonA.AgeOfDeath;
            int yearOfPersonB = _people.PersonB.YearOfDeath - _people.PersonB.AgeOfDeath;

            if (yearOfPersonA.IsZeroOrNegative() || yearOfPersonB.IsZeroOrNegative()) return -1;

            return Utils.GetAverage(GetNumberOfPeopleKilled(yearOfPersonA), GetNumberOfPeopleKilled(yearOfPersonB));
        }
    }
}
